shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,790 @@
+
+  while (stack.length > 0) {
+    const chosen = stack.pop();
+    let minimal = true;
+    for (const removedHashedClause in chosen) {
+      const newChosen = {};
+      for (const hashedClause in chosen) {
+        if (hashedClause !== removedHashedClause) {
+          newChosen[hashedClause] = chosen[hashedClause];
         }
-        if (!stackedUsers[sharer]) {
-          userStack.push(sharer);
-          stackedUsers[sharer] = true;
+      }
+
+      if (fullCover(newChosen)) {
+        minimal = false;
+        // TODO(perf): keep track of values we've already explored.
+        stack.push(newChosen);
+      }
+    }
+
+    if (minimal) {
+      const resultClause = new Clause();
+      const tokensUsed = {};
+      for (const hashedClause in chosen) {
+        resultClause.conjoin(chosen[hashedClause].clause);
+        for (const tokenId in chosen[hashedClause].tokensUsed) {
+          tokensUsed[tokenId] = true;
         }
       }
+
+      result[resultClause.hash()] = { clause: resultClause, tokensUsed: tokensUsed };
+    }
+  }
+
+  return result;
+}
+
+function proveClauses(db, context, goalClauses) {
+  // `goalClauses` is of the form of a result of `normalize()`, that is:
+  // `Map<ClauseId, {clause: Clause, tokensUsed: [tokenId]}>`.
+  //
+  // This function attempts to determine whether at least one of the input clauses holds.
+  // If a proof is found, returns `{yes: {tokensUsed: [tokenId]}}`.
+  // otherwise, returns {no: {}}
+  //
+  // The `db` parameter is optional. If it is present, the database will be queried as needed.
+  // If left out, the computation will proceed with only the tokens already present in `context`.
+
+  const clauses = {};
+  const clauseStack = []; // clause IDs
+
+  for (const hashedClause in goalClauses) {
+    const goalClause = goalClauses[hashedClause];
+    const tokensUsed = goalClause.tokensUsed;
+    if (goalClause.clause.isEmpty()) {
+      // We have an empty goal. Trivially true.
+      return { yes: { tokensUsed: tokensUsed } };
+    }
+
+    clauses[hashedClause] = { clause: goalClause.clause, tokensUsed: tokensUsed };
+    clauseStack.push(hashedClause);
+  }
+
+  const grainsFullyQueried = {};
+
+  while (clauseStack.length > 0) {
+    const hashedClause = clauseStack.pop();
+    const clause = new Clause();
+    clause.conjoin(clauses[hashedClause].clause);
+    const tokensUsed = clauses[hashedClause].tokensUsed;
+    const goal = clause.popFirstGoal();
+    const grainId = goal.vertex.grain._id;
+
+    if (db && !(grainId in grainsFullyQueried)) {
+      context.addGrain(db, grainId);
+    }
+
+    grainsFullyQueried[grainId] = true;
+
+    const viewInfo = context.grains[grainId].cachedViewInfo || {};
+    const result = backpropagateVertex(context, goal.vertex, goal.permissions, viewInfo);
+    const newGoals = normalize(result, goal.permissions);
+
+    // For each new clause, conjoin it with the remaining goals in `clause`.
```

Comment:
I'm saying that feels weird to me. Instead given `A ∧ B ∧ C` I would start by trying to prove A, B, and C separately. Expanding one item and then forming a new clause seems a lot harder to reason about.

What if there are cycles?

`A ∧ B ∧ C` -> `A ∧ W ∧ B ∧ C` -> `A ∧ W ∧ W ∧ B ∧ C` -- here the two W's would be de-duped of course, and then we're back at `A ∧ W ∧ B ∧ C`, which I guess would already be in the map and therefore we don't have a loop? But I'm having trouble being confident that there aren't other cases that become problematic, since this seems like an unusual way of thinking.